### PR TITLE
Modularize app-boot.

### PR DESCRIPTION
**Pull request by nathanhammond**
[Rendered](https://github.com/nathanhammond/ember-cli-rfcs/blob/master/active/0030-modularize-app-boot.md)

### CODE REVIEW

**Comment by nathanhammond** [comment]
Earlier discussion was here: https://github.com/ember-cli/ember-cli/issues/5242

File: active/0030-modularize-app-boot.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+This will reduce flexibility, but having opinions is a good thing.
+
+# Alternatives
+
+Leave this alone. After https://github.com/ember-cli/ember-cli/pull/5246 lands it isn't as pressing of an issue.
+
+# Unresolved questions
+
+* How do we make this play nicely for multiple apps on the same page?
```
Comment (nathanhammond) [inline active/0030-modularize-app-boot.md:36]: I believe this is resolved by specifying which modules to `require`.

**Comment by rwjblue** [comment]
Thanks, sorry for the run around.

**Comment by nathanhammond** [comment]
No problem, just so long as it doesn't go to RFCs to die. :stuck_out_tongue:

File: active/0030-modularize-app-boot.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+* [ember-cli](https://github.com/ember-cli/ember-cli/blob/a2a9a886e146ce7bdfbdd5338542538ea7b38d14/lib/broccoli/ember-app.js#L1600) - Generates the initial require.
+* [ember-cli-delay-app-boot](https://github.com/rwjblue/ember-cli-delay-app-boot/blob/master/index.js#L31) - Throws a timeout around the initial require and includes some copypasta from above.
+* [ember-cli-fastboot](https://github.com/tildeio/ember-cli-fastboot/blob/2058dd1ebf08be818851d1e65651db49cdd0e69f/index.js#L73) - This adds a new specially-named module and doesn't call `require`. Instead, it itself becomes the module which must be required to boot the app.
+- In the future (read: I'm already encountering this) a Service Worker may wish to tell the application when it should boot or bind that functionality onto a promise chain. My current approach is to delay responding to the app-boot request and maintain that promise chain inside of the service worker.
+
+# Detailed design
```
Comment (stefanpenner) [inline active/0030-modularize-app-boot.md:18]: after reading the detailed designed, i am confused what is actually being proposed.

**Comment by stefanpenner** [comment]
left a note, I do believe we should still explore this. As fast-boot and other scenarios would benefit.

At this time though, its not obvious to me what the detailed design is suggesting we do. Maybe some short code-examples could help me understand better.

**Comment by nathanhammond** [comment]
Goal: eliminate `app-boot` as an [appended piece of imperative code](https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/ember-app.js#L1192).

In short:
1. Make `app-boot` a module, [similar or identical to the way Fastboot does it](https://github.com/ember-fastboot/ember-cli-fastboot/blob/master/lib/utilities/fastboot-app-module.js).
2. Deprecate usage of the existing hook.
3. Create a config option to specify which module(s) gets `require`d and invoked (in order) to boot the app. Default is approximately `require('ember-app')()`.

I'm realizing that I didn't actually describe the API for specifying the configuration, but if we're okay on principle it's small enough that I can just implement it as a strawman PR that can land simultaneously with an improved RFC.

Thoughts @stefanpenner?

**Comment by rwjblue** [comment]
We are working on closing the ember-cli/rfcs repo in favor of using a single central RFC's repo for everything. This was laid out in https://emberjs.github.io/rfcs/0300-rfc-process-update.html. 

Sorry for the troubles, but would you mind reviewing to see if this is still something we need, and if so migrating this over to emberjs/rfcs?

Thank you!

**Pull request closed without merging**